ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java

```diff
@@ -1375,6 +1500,36 @@ protected void maybeProcessCallbacks() {
     @Override
     void handleBody(InputStream payload, MessageMetadata messageMetadata, MessageInfo messageInfo)
         throws IOException, MessageFormatException {
+
+      if (isMetadataCacheHit()) {
+        /**
+         * If we find blob metadata in frontend cache, validate it and use it.
+         */
+        BlobProperties receivedBlobProperties = MessageFormatRecord.deserializeBlobProperties(payload);
+        updateTtlIfRequired(serverBlobProperties, messageInfo);
+        if (serverBlobProperties.toString() != receivedBlobProperties.toString()) {
+          /**
+           * If we find blob properties is in our frontend cache, we request just blobInfo from backend to validate
+           * the cache entry. If cached blob properties does not match received blob properties,
+           * we must delete the cache entry and throw a retriable exception to the user.
+           * The next request will re-fill the cache with a valid entry if the blob stills exists.
+           * We do not need to decrypt anything at this point.
+           */
```

Comment:
I don't think so.

The blob property, except for ttl, is immutable. So if the blob property from server is different from cache, them it's because the ttl is different, which means client updated ttl before get. This just means that we have to update the ttl in the cache, we don't have to through an error back to client.

Reply:
I agree. I update the TTL and then compare the cached and received blob properties. This if-clause is for the rare case when something goes wrong.